File: csharp/LogicAndTrick.WikiCodeParser/Processors/INodeProcessor.cs
using System.Collections.Generic;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Processors
{
    public interface INodeProcessor
    {
        /// <summary>
        /// Higher priority processors are run first.
        /// </summary>
        int Priority { get; set; }

        /// <summary>
        /// Return true if the given node should be processed by this processor
        /// </summary>
        bool ShouldProcess(INode node, string scope);

        /// <summary>
        /// Process a node and return the nodes that will replace it.
        /// </summary>
        /// <returns>The nodes to replace the given node with</returns>
        IEnumerable<INode> Process(Parser parser, ParseData data, INode node, string scope);
    }
}

File: php/src/Processors/INodeProcessor.php
<?php

namespace LogicAndTrick\WikiCodeParser\Processors;

use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;

interface INodeProcessor
{
    /**
     * Higher priority processors are run first.
     */
    function Priority() : int;

    /**
     * Return true if the given node should be processed by this processor
     */
    function ShouldProcess(INode $node, string $scope) : bool;

    /**
     * Process a node and return the nodes that will replace it.
     * @return INode[]
     */
    function Process(Parser $parser, ParseData $data, INode $node, string $scope) : array;
}
